Use post_author instead of the post meta '_customer_user'

rodrigoprimo opened the pull request:
For orders, WC core uses the post meta `_customer_user` to store the customer ID. This PR stores the customer ID in the `post_author` field of the `wp_posts` as well (the `_customer_user` post meta is still populated to keep backwards compatibility). Currently the `post_author` field is not used by WC core when the post_type is `shop_order` (it sets it to 1 for all orders).

Having the customer ID in the `post_author` field should significantly improve the performance of queries that need to filter orders by customer ID (see an example [below](https://github.com/woocommerce/woocommerce/pull/17895#issuecomment-350712859)).

~Before merging this we want to make sure it won't break extensions that might be using post_author field for orders.~ - Already checked by posting on the WC dev blog and asking about in the WC dev chat.

Related #17087

rodrigoprimo committed:
First pass at using post_author instead of the post meta '_customer_user'

mikejolley commented on includes/data-stores/abstract-wc-order-data-store-cpt.php:
```diff
@@ -69,7 +69,7 @@ public function create( &$order ) {
 					'post_status'   => 'wc-' . ( $order->get_status( 'edit' ) ? $order->get_status( 'edit' ) : apply_filters( 'woocommerce_default_order_status', 'pending' ) ),
 					'ping_status'   => 'closed',
-					'post_author'   => 1,
+					'post_author'   => is_callable( array( $order, 'get_customer_id' ) ) ? $order->get_customer_id() : 1,
```
The default "1" corresponds to admin user. Should we try to force 0 for guest orders?

rodrigoprimo replied:
@mikejolley after this change, `1` will be used only for refund orders which it seems to me that is fine since refunds don't have a `_customer_user` post meta. `0` will be used for regular orders placed by guest users. Am I missing something?

mikejolley replied:
I think we need to keep it consistent throughout order types and always set it to the user who owns the order/refund? I worry that using 1 could mean someone using this field would get extra rules when getting admin (user 1) orders.

mikejolley commented on includes/data-stores/abstract-wc-order-data-store-cpt.php:
```diff
@@ -146,6 +146,7 @@ public function update( &$order ) {
 				'post_excerpt'      => $this->get_post_excerpt( $order ),
 				'post_modified'     => isset( $changes['date_modified'] ) ? gmdate( 'Y-m-d H:i:s', $order->get_date_modified( 'edit' )->getOffsetTimestamp() ) : current_time( 'mysql' ),
 				'post_modified_gmt' => isset( $changes['date_modified'] ) ? gmdate( 'Y-m-d H:i:s', $order->get_date_modified( 'edit' )->getTimestamp() ) : current_time( 'mysql', 1 ),
+				'post_author'       => is_callable( array( $order, 'get_customer_id' ) ) ? $order->get_customer_id() : 1,
```
1

mikejolley commented on includes/wc-update-functions.php:
```diff
@@ -1479,6 +1479,21 @@ function wc_update_330_image_options() {
+/**
+ * Copy order customer_id from post meta to post_author.
+ */
+function wc_update_330_order_customer_id() {
```
If we adjust the 1 -> 0, we may need to adjust more than these orders. Although, 0 might be used already depending on _customer_user.

rodrigoprimo committed:
Merge branch 'master' into update/use-post-author-for-customer-id

rodrigoprimo committed:
Set post_author to 0 for refund orders

rodrigoprimo commented:
@mikejolley, on commit ad43044 I changed the code to set post_author to `0` for refund orders as you suggested. Now I'm downloading to my local machine a big WooCommerce database to change a few queries to use `post_author` instead of `_customer_user` and test the impact on the performance.

rodrigoprimo committed:
Make sure post_author is used only after WC 3.3.0 DB upgrade routine is executed

rodrigoprimo committed:
Use post_author field instead of post meta in orders admin page

This commit changes the query used to get a list of orders in the orders admin page to use post_author field instead of the _customer_user post meta when filtering orders by customer.

Testing this change on a local WooCommerce install with about 2.5 million entries in the wp_posts table and about 100 million entries in the wp_postmeta table I got the following results:

- The original query using _customer_user took 28 seconds to run
- The new query using post_author took 0.0008 seconds to run

rodrigoprimo commented:
On commit 6ecc3bc, I changed the query used to get a list of orders in the orders admin page to use `post_author` field instead of `_customer_user` post meta when filtering orders by customer.

Testing this change on a local WooCommerce install with about 2.5 million entries in the wp_posts table and about 100 million entries in the wp_postmeta table I got the following results:

- The original query using `_customer_user` took 28 seconds to run
- The new query using `post_author` took 0.0008 seconds to run

I also noticed that the queries executed on the update function (`wc_update_330_order_customer_id()`) are somewhat slow. It took about 12 minutes to run this function on the same test database. We migth need to improve those queries before merging this PR if we decide to move forward with the change proposed here.

rodrigoprimo committed:
Merge branch 'master' into update/use-post-author-for-customer-id

rodrigoprimo committed:
Merge branch 'master' into update/use-post-author-for-customer-id

rodrigoprimo committed:
phpcs fixes

rodrigoprimo committed:
Bump WC version to 3.4.0

thenbrent commented:
Exciting to see this change.

Could the `post_type` used in `wc_update_340_order_customer_id()` be passed through a filter (and switched to use an `IN()` clause to support multiple post types)?

The `shop_subscription` post type in Subscriptions uses `'post_author'` the same way as the `shop_order` post type uses it. I'd like to continue that consistency, so I'd like to be able to add the `shop_subscription` post type to the update query and update those posts at the same time.

One potential issue too - the query in `wc_update_340_order_customer_id()` could be updating 100,000s rows at once. Many years ago, I had queries that changed a lot of data like that fail on cheap shared hosts. It might be worth testing this update query on a couple of $5 / month shared hosts with a large set of orders to make sure it doesn't choke completely.

franticpsyx commented:
> One potential issue too - the query in `wc_update_340_order_customer_id()` could be updating 100,000s rows at once. Many years ago, I had queries that changed a lot of data like that fail on cheap shared hosts. It might be worth testing this update query on a couple of $5 / month shared hosts with a large set of orders to make sure it doesn't choke.

+1

Might make sense to update the DB in batches -- for inspiration see https://github.com/somewherewarm/woocommerce-product-bundles/blob/master/includes/wc-pb-update-functions.php#L147

rodrigoprimo committed:
Add filter to list of post_types used to build the query to copy _customer_user to post_author

rodrigoprimo commented:
Thanks for your input @thenbrent and @franticpsyx.

> Could the post_type used in wc_update_340_order_customer_id() be passed through a filter (and switched to use an IN() clause to support multiple post types)?

Filter added in commit 1f39f7d (@claudiosanches, do you see a way to build [this](https://github.com/woocommerce/woocommerce/pull/17895/commits/1f39f7d63378caa86002795f19d2dae8890dc971#diff-4765824e656c3231e5109392672d7b72R1555) query without having to use `phpcs:ignore WordPress.WP.PreparedSQL.NotPrepared`?)

> It might be worth testing this update query on a couple of $5 / month shared hosts with a large set of orders to make sure it doesn't choke.

It is on my list to check for ways to reduce the impact of the queries executed in `wc_update_340_order_customer_id()`. I'm just waiting a few more days to hear the feedback from the community before resuming the work on this PR. Thanks for pointing me to woocommerce-product-bundles code to update the DB in batches, @franticpsyx.

rodrigoprimo committed:
Merge branch 'master' into update/use-post-author-for-customer-id

kovshenin commented:
Thanks for your hard work on this @rodrigoprimo! I have a few thoughts/questions.

Have you considered the possible security implications behind this change? Some logic around meta capabilities in WordPress core makes assumptions, that if a post is not published or scheduled, then its author is allowed to edit it: https://github.com/WordPress/WordPress/blob/master/wp-includes/capabilities.php#L160 So I want to make sure that by setting the `post_author` on an order to the customer, we're not granting that customer `edit_post` capabilities for that order.

When selecting all the customer IDs I would probably avoid joining on the wp_posts table. The vast majority of WooCommerce installs will already have the customer_id only on `shop_order` types. It's okay if a very small subset of the data doesn't fit our post type, we can filter it out later, when running the updates.

Updating is likely the slowest part, and it's going to be a problem with high-volume stores. I like @franticpsyx's mention of batching, though I can't access that link. WordPress core does this nicely with `_split_shared_term()` utilizing cron jobs: https://github.com/WordPress/WordPress/blob/master/wp-includes/taxonomy.php#L3682 Also note how core cleans object cache entries which are affected by these changes, we should do the same with `clean_post_cache()` or `wp_cache_delete()`. This also means that we need to look for a batch size that will work on 64M memory limit installs, because `get_post()` will consume memory, `$wpdb->update()` will consume memory (if SAVEQUERIES is on), etc.

It's not going to be the fastest way, but for the majority of stores I think it'll be okay. For high-volume stores we can provide some guidance with some snippets of how to do it faster. For example on my test store with about 520k orders and 320k users I was able to perform the update in about 15 seconds using a temporary table:

```
> CREATE TEMPORARY TABLE customers_map (post_id BIGINT(20), customer_id BIGINT(20), PRIMARY KEY(post_id));
```

Then insert all existing `_customer_user` records into this table:

```
> INSERT IGNORE INTO customers_map (SELECT post_id, meta_value FROM wp_postmeta WHERE meta_key = '_customer_user');
Query OK, 518480 rows affected, 2 warnings (3.54 sec)
```

And finally update `wp_posts` by joining on this table (you may need to turn off `SQL_SAFE_UPDATES` to run this query):

```
> UPDATE wp_posts JOIN customers_map ON wp_posts.ID = customers_map.post_id SET wp_posts.post_author = customers_map.customer_id WHERE post_type = 'shop_order';
Query OK, 518359 rows affected, 1 warning (11.47 sec)
```

Note that all the affected orders will be locked for the duration of the query. In a MyISAM environment the whole table will be locked. If this is undesirable, then the update can also be performed in batches. Then the only thing remaining would be to clean out those 500k keys from Memcached and reindex the order data in Elasticsearch.

rodrigoprimo commented:
Thanks for your input, @kovshenin!

> Have you considered the possible security implications behind this change? Some logic around meta capabilities in WordPress core makes assumptions, that if a post is not published or scheduled, then its author is allowed to edit it: https://github.com/WordPress/WordPress/blob/master/wp-includes/capabilities.php#L160 So I want to make sure that by setting the post_author on an order to the customer, we're not granting that customer edit_post capabilities for that order.

That is a good point. I had considered this when I started working on this PR, but it is always good to double check my assumptions to make sure this change will no introduce any security issues. The `shop_order` CPT has its own set of capabilities ([includes/class-wc-post-types.php#L390](https://github.com/woocommerce/woocommerce/blob/0599c06b20a6fd38547588d25f2e3c43a6d4503a/includes/class-wc-post-types.php#L390)) and, as far as I can see, using `post_author` to store the customer ID would be an issue only if `shop_order` used the default WP capabilities. WC assigns its custom capabilities only to the roles `administrator` and `shop_manager` ([includes/class-wc-install.php#L891](https://github.com/woocommerce/woocommerce/blob/a7d2c5dd6e7559348e2f0c15b8222e506f3c31d9/includes/class-wc-install.php#L891)).

> Then the only thing remaining would be to clean out those 500k keys from Memcached and reindex the order data in Elasticsearch.

Not sure if I understand what you meant here. Could you please elaborate? Cleaning the cache as you mentioned above is not enough to clean Memcached and/or reindex Elasticsearch for sites using them?

I will work on a new version of the function to migrate data based on your suggestions and get back to you. Thanks again!

rodrigoprimo committed:
Merge branch 'master' into update/use-post-author-for-customer-id

rodrigoprimo committed:
Reposition wc_update_340_order_customer_id() function

So that it is listed together with WC 3.3 update functions

rodrigoprimo committed:
Change WC_Background_Updater to allow in-task resource monitoring

This change makes it possible to monitor resource usage inside each of the database update functions. Useful when doing resource intensive updates to stop the execution and requeue before the execution is aborted.

rodrigoprimo committed:
Migrate _customer_user to post_author in batches

This will make the update slower but safer when running in cheap servers.

rodrigoprimo committed:
Punt to WC version 3.5.0

rodrigoprimo committed:
Migrate data to post_author in a single go when using WP-CLI

rodrigoprimo committed:
Monitor only memory limit while running the data migration

Turns out WP_Background_Process uses an arbitrary value of 20 seconds of wall clock time to measure PHP "time limit" (https://github.com/woocommerce/woocommerce/blob/b0617a13c4ddba13559eaac1f295c9622b333523/includes/libraries/wp-background-process.php#L385) instead of cpu time which apparently it is not easy to measure in PHP. Since system calls (like database calls) are not included in the PHP maximum execution time, 20 seconds of wall clock is often way less than PHP time limit.

Thus, this commit removes the call to WP_Background_Process::time_exceeded() while running the function to migrate the post meta _customer_user to the field post_author. If the script times out, WP_Background_Process will restart it.

kovshenin commented:
> Not sure if I understand what you meant here. Could you please elaborate? Cleaning the cache as you mentioned above is not enough to clean Memcached and/or reindex Elasticsearch for sites using them?

Sorry, what I meant was that after the SQL queries that I ran (not the batched approach), some data in Memcached and Elastic could be stale. If you're doing batches via WP APIs and not direct database calls, using `clean_post_cache()` should take care of Memcached, and potentially any ES integrations.

rodrigoprimo committed:
Cache query to get the list of orders placed by the user ID 1

rodrigoprimo committed:
Set post_author to 0 on all orders that belong to a deleted customer

rodrigoprimo committed:
Fix PHPCS violations in includes/admin/reports/class-wc-report-customers.php

rodrigoprimo committed:
Update wc_customer_bought_product() and customers report to use post_author instead of _customer_user

rodrigoprimo committed:
Update WC API v1 and v2 to use post_author instead of _customer_user

rodrigoprimo committed:
Update WC_Customer_Data_Store to use post_author instead of _customer_user

rodrigoprimo committed:
Update WC_Order_Data_Store_CPT::get_wp_query_args() to use post_author instead of _customer_user

rodrigoprimo committed:
Merge branch 'master' into update/use-post-author-for-customer-id

rodrigoprimo commented:
A few weeks ago I finished a new version of the function that migrates data from `_customer_user` to `post_author`. Now the function will do the migration in small batches if called from the wp-admin and will migrate everything at once if called from the command line.

This week I finished this PR by updating all the places where WC core was using `_customer_user` to use `post_author` instead.

This PR is now ready for review.

mikejolley commented on includes/wc-update-functions.php:
```diff
@@ -1773,3 +1773,123 @@ function wc_update_340_last_active() {
+		}
+
+		// Set post_author to 0 instead of 1 on all shop_order_refunds.
+		while ( true ) {
```
This could be split into a separate callback - then background processing will take care of queuing it after orders are done.

rodrigoprimo replied:
Could you elaborate on why you are suggesting this? It seems to me that it is better to keep all database changes needed for this PR in a single function.

mikejolley replied:
@rodrigoprimo You have several while loops in sequence. And you're handling memory usage manually.

If `wc_update_350_order_customer_id` was split into multiple callbacks, e.g. `wc_update_350_order_customer_id_refunds` etc, it will call them in sequence and manage memory for you.

Does that make sense?

rodrigoprimo replied:
Even if I split into two callbacks, I would still want to manage memory usage manually inside each while loop so I don't see any advantage to split.

mikejolley commented on includes/wc-update-functions.php:
```diff
@@ -1773,3 +1773,123 @@ function wc_update_340_last_active() {
+		$query = $wpdb->prepare(
+			// phpcs:disable WordPress.WP.PreparedSQL.NotPrepared
+			"SELECT ID FROM {$wpdb->posts}
+	        WHERE post_author = 1 AND post_type IN ({$post_types_placeholders}) $admin_orders_sql
```
indenting is odd here.

rodrigoprimo replied:
Fixed in 397e40f1659e38627be916c6f0114b33d0c2cf2c

mikejolley commented on includes/wc-update-functions.php:
```diff
@@ -1773,3 +1773,123 @@ function wc_update_340_last_active() {
+			);
+
+			// Cache the list of orders placed by the user ID 1 as to large stores this query can be slow.
+			set_transient( 'wc_update_350_admin_orders', $admin_orders, DAY_IN_SECONDS );
```
Is this more efficient than a routine to first set all post author fields to 0 before doing this?

rodrigoprimo replied:
I remember I considered updating all post author fields to 0 but to be honest I don't remember why I chose another approach. One thing to consider is that in the when running the update via the wp-admin I'm not trying to be as efficient as possible but instead to be mindful of resources consumption. Updating all post_author fields of all orders at once is likely to consume more resources than getting the list of orders that belong to the admin user.

mikejolley replied:
Even if thats a pure sql thing? The method you're using involves PHP loops, and if we're missing something (size of admin orders) it could cause resource issues itself.

rodrigoprimo replied:
After running some tests, I now remember why I decided to get a list of orders that belong to the admin instead of updating all post_author fields to 0. It seems to me that updating all post_author fields to 0 will lock wp_posts to inserts and updates while the query is running which can be problematic for large and busy stores as the query will take some time to complete while the site is live. Queries could timeout while waiting for a lock. The same wouldn't happen with the select query to get all the orders placed by the admin.

It is important to mention that I don't have a deep understanding of how InnoDB locks works, so there is a chance I'm missing something.

mikejolley reviewed (changes requested):
Added some comments and questions inline :)

Does this impact our (recent) lookup table idea?

rodrigoprimo committed:
Merge branch 'master' into update/use-post-author-for-customer-id

rodrigoprimo committed:
Fix some code untitentional code changes added during the last merge from master

rodrigoprimo commented on includes/class-wc-background-updater.php:
```diff
@@ -99,18 +99,26 @@ protected function task( $callback ) {
 			$logger->info( sprintf( 'Running %s callback', $callback ), array( 'source' => 'wc_db_updates' ) );
-			call_user_func( $callback );
-			$logger->info( sprintf( 'Finished %s callback', $callback ), array( 'source' => 'wc_db_updates' ) );
+			$result = call_user_func( $callback, $this );
```
Done :)

rodrigoprimo committed:
Add WC 3.5.0 DB version update callback

rodrigoprimo commented on includes/class-wc-install.php:
```diff
@@ -104,6 +104,9 @@ class WC_Install {
 			'wc_update_340_db_version',
 		),
+		'3.5.0' => array(
+			'wc_update_350_order_customer_id',
```
Good catch, thanks.

rodrigoprimo committed:
Pass user_id via $wpdb->prepare()

rodrigoprimo commented on includes/wc-user-functions.php:
```diff
@@ -507,6 +523,21 @@ function wc_reset_order_customer_id_on_deleted_user( $user_id ) {
+	$wpdb->query(
+		// phpcs:disable WordPress.WP.PreparedSQL.NotPrepared, WordPress.DB.PreparedSQLPlaceholders.UnfinishedPrepare
+		$wpdb->prepare(
+			"UPDATE {$wpdb->posts} SET `post_author` = 0 WHERE post_type IN ({$post_types_placeholders}) AND post_author = {$user_id}",
```
I forgot to pass user_id via prepare. This is now fixed in 77add4670573b79634c4b192a01932a1a7569eb6

rodrigoprimo committed:
Fix indentation

rodrigoprimo commented on includes/wc-update-functions.php:
```diff
@@ -1773,3 +1773,123 @@ function wc_update_340_last_active() {
+		}
+
+		// Query to get a batch of orders IDs to change.
+		$query = $wpdb->prepare(
```
Prepare is being used. `$post_types_placeholders` contains a list of placeholders and `$post_types` the actual values. This is needed as the number of `$post_types` can vary if the filter `woocommerce_update_350_order_customer_id_post_types` is used. Did you mean something else that I missed?

mikejolley replied:
Nope, missed that generation. I'd usually just run esc_sql over the post types and avoid prepare in that case, but I guess this is just as valid and avoids a phpcs comment. 👍

rodrigoprimo commented:
Thanks for your comments and questions, @mikejolley. I fixed all the issues you found and this PR is ready for review again.

> Does this impact our (recent) lookup table idea?

@kloon is probably the best person to answer this question as he suggested the lookup tables. Generally speaking, my understanding is that there is some overlap between this PR and the lookup tables, but even if we create those tables having the customer ID in the post_author field is still valid. The more we can use WP tables schema the better as we can benefit from WP built-in functions.

kloon commented:
> Generally speaking, my understanding is that there is some overlap between this PR and the lookup tables, but even if we create those tables having the customer ID in the post_author field is still valid

That is correct, even though there is some overlap the lookup tables solves slow lookups in general across different scenarios and not just order <> user lookups.

mikejolley commented on includes/wc-update-functions.php:
```diff
@@ -1773,3 +1773,123 @@ function wc_update_340_last_active() {
+	$post_types              = (array) apply_filters( 'woocommerce_update_350_order_customer_id_post_types', array( 'shop_order' ) );
+	$post_types_placeholders = implode( ', ', array_fill( 0, count( $post_types ), '%s' ) );
+
+	if ( defined( 'WP_CLI' ) && WP_CLI ) {
```
Yeah, just worried how we roll that out safely :)

rodrigoprimo replied:
I just run the migration again via wp-admin on a store with 1,900,000 orders (2,600,000 entries in wp_posts and 101,000,000 entries in wp_postmeta) running on my local environment. It took two hours and twenty-two minutes for the update to complete. Everything worked as expected.

mikejolley replied:
What state was the site in during that time?

mikejolley replied:
What state was the site in during that time?

rodrigoprimo replied:
The site was responsive while the update was running but that is a local environment so no real traffic.

rodrigoprimo committed:
Merge branch 'master' into update/use-post-author-for-customer-id

rodrigoprimo committed:
Add @return for wc_update_350_order_customer_id() docblock

mikejolley committed:
Bump version so update routine gets triggered in master

rodrigoprimo commented:
@claudiulodro, happy if we merge this PR and ship it in 3.5?

claudiulodro commented:
@rodrigoprimo Sounds good. It should make some of the v3 API things easier also. Thanks!

stianandreassen commented:
The `wc_update_350_order_customer_id()` function broke several orders on a site I maintain. 

Every order that had been placed manually in the admin by a shop manager got changed, so the **shop manager was set to customer on the order.** 

On lines 1901–1914 of **wc-update-functions.php** there seems to be a reference to user ID 1: 
`// Get the list of orders that we don't want to change as they belong to user ID 1.`
`[…]  AND meta_key = '_customer_user' AND meta_value = 1`. 

Seems strange to only bypass orders from user 1? Shouldn’t this include all shop managers if `$order['created_via'] == 'admin'`?

kloon commented:
@stianandreassen could you perhaps clarify how you are relying on this author field? Since only orders placed via wp-admin would have the author set to the logged in person who created it, all orders placed via the frontend prior to WooCommerce 3.5 would have an ID of 1.

stianandreassen commented:
@kloon I might have missed the mark with the reference to the lines in **wc-update-functions.php**

I just tested on a clean install of WP, with **only** WooCommerce 3.4.6 and Storefront 2.3.3 activated, and got the same problem. I create and order as a shop manager in v3.4.6, then upgrade to 3.5.0, and the customer is replaced with the shop manager. 

Before update (v3.4.6):
https://www.dropbox.com/s/oao4r4kpgats270/woo_350_1.png?dl=0

After update (v3.5.0):
https://www.dropbox.com/s/6etozrvgj8k64es/woo_350_2.png?dl=0

rodrigoprimo commented:
@stianandreassen thanks for your comments. I can confirm the problem that you reported. I'm creating an issue detailing what I have found so far and will start working on a solution soon afterward.

rodrigoprimo commented:
See #21656 for my understanding of this problem and some first thoughts on how I'm planning to address it.

crstauf commented:
@rodrigoprimo Is this being reconsidered for 3.6?

kloon commented:
@crstauf there are plans to revisit this soon, but not for the 3.6 release.

crstauf commented:
@kloon @rodrigoprimo Is there a current issue/PR for the reintroduction of this? I've just bounced around a few, and none of them seem current.

rodrigoprimo commented:
@crstauf there is not an issue or PR open for the reintroduction of this. This is not something that we are currently working on.
